controllers/controlplane_controller_reconciler_utils.go

```
@@ -267,3 +267,46 @@ func (r *ControlPlaneReconciler) ensureCertificate(
 
 	return created, secretName, err
 }
+
+// garbageCollectClusterResources deletes all the cluster wide resources
+// owned by deleted controlplane.
+func (r *ControlPlaneReconciler) garbageCollectClusterResources(
+	ctx context.Context,
+	controlplane *operatorv1alpha1.ControlPlane,
+) error {
+	// remove all owned ClusterRoleBindings.
+	clusterRoleBindings, err := k8sutils.ListClusterRoleBindingsForOwner(
+		ctx, r.Client,
+		consts.GatewayOperatorControlledLabel, consts.ControlPlaneManagedLabelValue, controlplane.UID,
+	)
+	if err != nil {
+		return err
+	}
+	for i := range clusterRoleBindings {
+		err = r.Client.Delete(ctx, &clusterRoleBindings[i])
+		if err != nil {
+			return err
+		}
```

I think we don't want to return here. The reason for this is that (if I understand this right) we want to remove as much as possible, not "fail fast" during removal.

So perhaps we can use something like https://github.com/hashicorp/go-multierror or an alternative and just group errors but at the same time continue execution to remove as much as possible?

---

updated.